### PR TITLE
Better error when submitting order with out of stock product

### PR DESCRIPTION
**Pull request by marlena-b**
Issue: https://github.com/RailsEventStore/ecommerce/issues/375

Before when a product was added to an order and subsequently went out of stock before submission, the system incorrectly showed a success message  "Your order is being submitted" while leaving the order in draft status. Now it shows "Order can not be submitted! Product not available in requested quantity!". 
Zrzut ekranu 2024-09-1 o 22 05 50

There are a lot of changes. I will try to explain all of them in the comments below. 

The implementation for the Result object was taken from this presentation: https://www.icloud.com/keynote/0x2bfAdGek3Eg-I_2kIomD-6g#Controller_refactoring (it lacks notes but you can check slides 7 and 17 for "before" and "after")

### CODE REVIEW

File: ecommerce/crm/lib/crm/order.rb [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
     include AggregateRoot
-    CustomerAlreadyAssigned = Class.new(StandardError)
 
     def initialize(id)
       @id = id
     end
 
     def set_customer(customer_id)
-      raise CustomerAlreadyAssigned if @customer_id
+      return if @customer_id == customer_id
```
Comment (marlena-b) [inline ecommerce/crm/lib/crm/order.rb:10]: I think we need to allow users to reassign customers. If order submission fails the first time the user is still being assigned. We then display the form with an error again and the user can change the customer.

File: ecommerce/processes/lib/processes/reservation_process.rb [hunk 2]
```diff
@@ -28,10 +21,29 @@ def call(event)
     private
 
     def reserve_stock(state)
+      unavailable_products = []
       state.order_lines.each do |product_id, quantity|
         command_bus.(Inventory::Reserve.new(product_id: product_id, quantity: quantity))
         state.product_reserved(product_id)
+      rescue Inventory::InventoryEntry::InventoryNotAvailable
+        unavailable_products << product_id
+      end
```
Comment (marlena-b) [inline ecommerce/processes/lib/processes/reservation_process.rb:30]: ReservationProcess no longer stops at the first unavailable product. Instead, it tries to reserve all of them and records which were unavailable. This way we can later display the full list of unavailable products.

File: ecommerce/processes/lib/processes/reservation_process.rb [hunk 3]
```diff
@@ -28,10 +21,29 @@ def call(event)
...
+
+      if unavailable_products.empty?
+        event = ReservationProcessSuceeded.new(data: { order_id: state.order_id })
+      else
+        release_stock(state)
+        event = ReservationProcessFailed.new(data: { order_id: state.order_id, unavailable_products: unavailable_products })
       end
+      event_store.publish(event, stream_name: stream_name(state.order_id))
```
Comment (marlena-b) [inline ecommerce/processes/lib/processes/reservation_process.rb:38]: ReservationProcess now emits events. We then subscribe to these events in the controller and display proper errors to the user.

File: ecommerce/processes/lib/processes/reservation_process.rb [hunk 4]
```diff
@@ -28,10 +21,29 @@ def call(event)
...
+    end
+
+    def order_side_effects(state)
+      event_store
+      .within { yield }
+      .subscribe(to: ReservationProcessFailed) { reject_order(state) }
+      .subscribe(to: ReservationProcessSuceeded) { accept_order(state) }
+      .call
```
Comment (marlena-b) [inline ecommerce/processes/lib/processes/reservation_process.rb:46]: I think these side effects could be extracted to fully-fledged event handlers. Then the ReservationProcess could have only a single responsibility. This PR is already big so I didn't want to extract it here. Let me know what you think and I will add an issue for this refactor.

File: rails_application/app/controllers/client/orders_controller.rb [hunk 5]
```diff
@@ -12,13 +12,12 @@ def new
     end
 
     def create
-      ActiveRecord::Base.transaction do
-        command_bus.(Ordering::SubmitOrder.new(order_id: params[:order_id]))
-        command_bus.(Crm::AssignCustomerToOrder.new(customer_id: cookies[:client_id], order_id: params[:order_id]))
+      Orders::SubmitService.new(order_id: params[:order_id], customer_id: cookies[:client_id]).call.then do |result|
```
Comment (marlena-b) [inline rails_application/app/controllers/client/orders_controller.rb:15]: The code in this controller got more and more complex so I extracted it to a service. 
I used a very simple Service object implementation combined with a simple but flexible Result object. This results in an expressive DSL like the one below :)

File: rails_application/app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -74,12 +74,13 @@ def remove_item
   end
 
   def create
-    ApplicationRecord.transaction { submit_order(params[:order_id], params[:customer_id]) }
-    redirect_to order_path(params[:order_id]), notice: "Your order is being submitted"
-  rescue Ordering::Order::IsEmpty
-    redirect_to edit_order_path(params[:order_id]), alert: "You can't submit an empty order"
-  rescue Crm::Customer::NotExists
-    redirect_to order_path(params[:order_id]), alert: "Order can not be submitted! Customer does not exist."
+    Orders::SubmitService.new(order_id: params[:order_id], customer_id: params[:customer_id]).call.then do |result|
```
Comment (marlena-b) [inline rails_application/app/controllers/orders_controller.rb:77]: I initially wanted to have two separate for the admin and client side but then they look almost entirely the same. So I decided to merge them so we don't duplicate too much code and tests.

File: rails_application/app/services/application_service.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+class ApplicationService
```
Comment (marlena-b) [inline rails_application/app/services/application_service.rb:1]: This is the first time service objects are used in this repo so I wanted to keep them as simple as possible. In my implementation service objects should only implement `initialize` and `call` methods and then we can call them like `MyService.call(..)`. This makes it easier to mock them (bc we only need to mock the class method `call`).

Please let me know if you prefer to use some other implementation instead, this is just my proposal.

File: rails_application/app/services/orders/submit_service.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+      if success
+        Result.new(:success)
+      else
+        Result.new(:products_out_of_stock, unavailable_products)
+      end
+      rescue Ordering::Order::IsEmpty
+        Result.new(:order_is_empty)
+      rescue Crm::Customer::NotExists
+        Result.new(:customer_not_exists)
```
Comment (marlena-b) [inline rails_application/app/services/orders/submit_service.rb:28]: `call` method returns a `Result` object which represents a logical path of execution combined with optional arguments (e.g. `unavailable_products`).

File: rails_application/app/services/result.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(status, *args)
+    @status = status
+    @args = args
+  end
+
+  def path(name, &block)
+    return unless @status == name.to_sym
+
+    block.call(*@args)
+  end
```
Comment (marlena-b) [inline rails_application/app/services/result.rb:13]: There is also an alternative implementation which uses more metaprogramming but results in a nicer DSL:

```
def method_missing(name, *_, &block)
  return unless @status == name.to_sym

  block.call(*@args)
end
```
and then the DSL looks like this:
```
MyService.call(params).then do |on|
  on.success { puts "Awesome!" }
  on.validation_failed { |errors| puts "Oh no, there are some errors: #{errors}" }
  on.something_else { |arg1, arg2| puts arg1 * arg2 }
end
```
Let me know which one you prefer :)

**Comment by andrzejkrzywda** [comment]
There are many changes in this pull request which I would like to debate separately, but as a whole I think it's ready to be merged.

1. Let's change the Result pattern to exceptions - we already use exceptions here a lot and we could save one pattern
2. Let's separate the service objects even if they are identical
3. No need for inheritance in service objects
4. process publishing events - let's have it for now, we will see how it feels in the code in the longer run, might be refactored later

**Pull request merged**